docs: add field decorators to undecorated classes migration (#33069)
angular/angular: aio/content/guide/migration-undecorated-classes.md

This migration adds an empty `@Directive()` decorator to undecorated

> base classes that are extended by either directives or components.

Reviewer: ```suggestion
base classes that use Angular features and are extended by directives or components.

For example, in the diff below, a `@Directive` decorator is added to `BaseMenu` because `BaseMenu` uses dependency injection.
```

> The schematic also copies any inherited directive or component metadata to the derived class.

Reviewer: ```suggestion
In the event that a directive or component is extended by a class without a decorator, the schematic copies any inherited directive or component metadata to the derived class.
```

export class SettingsMenu extends BaseMenu {}
```

> This schematic also works with Angular field decorators, including:

Reviewer: ```suggestion
This schematic also decorates classes that use Angular field decorators, including:
```

> This schematic also works with Angular field decorators that include:

Reviewer 2: ```suggestion
This schematic also works with Angular field decorators, including:
```

- `@Input()`
- `@Output()`
- `@HostBinding()`
- `@HostListener()`
- `@ViewChild()` / `@ViewChildren()`
- `@ContentChild()` / `@ContentChildren()`

> It does not include parameter decorators for dependency injection, such as `@Host()`, `@SkipSelf()`, `@Optional()`, and `@Self()`.

Reviewer 2: ```suggestion
This migration does not include _parameter_ decorators for dependency injection, such as `@Host()`, `@SkipSelf()`, `@Optional()`, and `@Self()`.
```

**Before:**
```ts
class Base {
  @Output()
  countChanged = new EventEmitter<number>();
}

@Directive({
  selector: '[myDir]'
})
class Dir extends Base {
}
```

**After:**
```ts
@Directive() // schematic adds @Directive()
class Base {
  @Output()

>   total: number;

Reviewer: Looks like the output code no longer matches the "before"

```suggestion
  countChanged = new EventEmitter<number>();
```

}

> @Directive()

Reviewer: ```suggestion
@Directive({
  selector: '[myDir]'
})
```

class Dir extends Base {
}
```

## Why is this migration necessary?

> When a class has a `@Directive()` or `@Component()` decorator,

Reviewer: It seems like we might need a subheader to organize this section

```suggestion
### Migrating Classes that Use DI

When a class has a `@Directive()` or `@Component()` decorator,
```

- `@ContentChild()` / `@ContentChildren()`

This migration does not include _parameter_ decorators for dependency injection, such as `@Host()`, `@SkipSelf()`, `@Optional()`, and `@Self()`.

**Before:**
```ts
class Base {
  @Output()
  countChanged = new EventEmitter<number>();
}

@Directive({
  selector: '[myDir]'
})
class Dir extends Base {
}
```

**After:**
```ts
@Directive() // schematic adds @Directive()
class Base {
  @Output()
  countChanged = new EventEmitter<number>();
}

@Directive({
  selector: '[myDir]'
})
class Dir extends Base {
}
```

## Why is this migration necessary?

### Migrating classes that use DI

When a class has a `@Directive()` or `@Component()` decorator,
the Angular compiler generates extra code to inject dependencies into
the constructor. When using inheritance, Ivy needs both the parent class
and the child class to apply a decorator to generate the correct code.

You can think of this change as two cases: a parent class is missing a
decorator or a child class is missing a decorator. In both scenarios,
Angular's runtime needs additional information from the compiler.
This additional information comes from adding decorators.

> #### Benefits of requiring a class decorator

Reviewer: This section appears to be in the wrong place now. It should be at the bottom of `Migrating Classes that Use Field Decorators`, as this section explains why we are now requiring them.

Always requiring a class decorator leads to two main benefits for Angular:

1. The previous behavior was inconsistent. Some Angular features required a decorator (dependency injection), but others did not. Now, all Angular features consistently require a class decorator.

1. Supporting undecorated classes increases the code size and complexity of Angular. Always requiring class decorators allows the framework to become smaller and simpler for all users.

### Migrating Classes that Use Field Decorators

In ViewEngine, base classes with field decorators like `@Input()` worked
even when the class did not have a `@Directive()` or `@Component()` decorator.
For example:

```ts
class Base {
  @Input()
  foo: string;
}

> @Directive()

Reviewer: Let's not use `@Directive()` unless we *specifically mean* an abstract directive. If we are just neglecting to write out the full decorator, we need a `...`

```suggestion
@Directive(...)
```

> Angular's run-time needs additional information from the compiler.

Reviewer: ```suggestion
 Angular's runtime needs additional information from the compiler.
```

This additional information comes from adding decorators.

> In ViewEngine, base classes with field decorators like `@Input()` worked

Reviewer: ```suggestion
### Migrating Classes that Use Field Decorators
In ViewEngine, base classes with field decorators like `@Input()` worked
```

even when the class did not have a `@Directive()` or `@Component()` decorator.
For example:

```ts
class Base {
  @Input()
  foo: string;
}

@Directive()
class Dir extends Base {
  ngOnChanges(): void {
    // notified when bindings to [foo] are updated
  }
}
```

However, this example won't compile with Ivy because the `Base` class
_requires_ either a `@Directive()` or `@Component()` decorator to generate
code for inputs, outputs, queries, and host bindings.

Always requiring a class decorator leads to two main benefits for Angular:

1. The previous behavior was inconsistent. Some Angular features required a decorator (dependency injection), but others did not. Now, all Angular features consistently require a class decorator.

1. Supporting undecorated classes increases the code size and complexity of Angular. Always requiring class decorators allows the framework to become smaller and simpler for all users.

> ### Decorator missing from parent class

Reviewer: This section seems to be in the wrong place now, given the additions. Move it into the ` Migrating Classes that Use DI` and make it into a `####`

Reviewer: This `Decorator missing from parent class` section still looks like it's in the wrong place. It should be inside `Migrating Classes that Use DI`. Seems like maybe you moved the section above it by mistake?

Angular's run-time needs additional information from the compiler.
This additional information comes from adding decorators.

In View Engine JIT and AOT, undecorated base classes, which

> have decorated fields such as`@Input()` or `@ViewChild()`, work.

Reviewer 2: ```markdown
In ViewEngine, base classes with field decorators like `@Input()` worked
even when the class did not have a `@Directive()` or `@Component()` decorator.
```

For example:

```ts
class Base {
  @Input()
  foo: string;
}

@Directive()
class Dir extends Base {
  ngOnChanges(): void {
    // notified when bindings to [foo] are updated
  }
}
```

However, in Ivy this model won't compile as there is no
decorator on `Base` which would generate an Ivy definition
to carry the information about inputs, outputs, queries, or
host bindings. Ivy always requires a `@Directive()`

> or `@Component()` selector for any class using Angular features.

Reviewer 2: ```markdown
However, this example won't compile with Ivy because the `Base` class
_requires_ either a `@Directive()` or `@Component()` decorator to generate
code for inputs, outputs, queries, and host bindings.
```

> There were two motivations for extending the requirement for a `@Directive()` decorator to include field decorators:

Reviewer 2: ```suggestion
Always requiring a class decorator leads to two main benefits for Angular:
```

> 1. The former distinction was arbitrary. Why does Dependency Injection (DI) require a decorator but other Angular features don't? It's hard to remember which is which. It's easier to just say "If you use any Angular feature, (whether it's a field decorator or DI), you must have a class decorator"

Reviewer 2: ```suggestion
1. The previous behavior was inconsistent. Some Angular features required a decorator (dependency injection), but others did not. Now, all Angular features consistently require a class decorator.
```

> 1. This change allows us to reduce the size of the framework code. Making Angular features work in undecorated base classes was complicated and expensive, code-wise. If we ask users to add the required decorators, the framework can be smaller for everyone.

Reviewer 2: ```suggestion
1. Supporting undecorated classes increases the code size and complexity of Angular. Always requiring class decorators allows the framework to become smaller and simpler for all users.
```

In the future, add `@Directive()` to base classes that
do not already have decorators and are extended by directives.

### Migrating classes that use field decorators

In ViewEngine, base classes with field decorators like `@Input()` worked
even when the class did not have a `@Directive()` or `@Component()` decorator.
For example:

```ts
class Base {
  @Input()
  foo: string;
}

@Directive(...)
class Dir extends Base {
  ngOnChanges(): void {
    // notified when bindings to [foo] are updated
  }
}
```

However, this example won't compile with Ivy because the `Base` class
_requires_ either a `@Directive()` or `@Component()` decorator to generate
code for inputs, outputs, queries, and host bindings.

> ### Benefits of requiring a class decorator

Reviewer: ```suggestion
```

Always requiring a class decorator leads to two main benefits for Angular:

1. The previous behavior was inconsistent. Some Angular features required a decorator (dependency injection), but others did not. Now, all Angular features consistently require a class decorator.

1. Supporting undecorated classes increases the code size and complexity of Angular. Always requiring class decorators allows the framework to become smaller and simpler for all users.

> ### Decorator missing from child class

Reviewer: This "Decorator missing from child class" section is still in the wrong place. It should be below "Decorating missing from parent class". These are the two cases described here: https://github.com/angular/angular/pull/33069/files#diff-1a7470dea32eb33d66ba772955bc07a3R112

## I'm a library author. Should I add the `@Directive()` decorator to base classes?

As support for selectorless decorators is introduced in
Angular version 9, if you want to support Angular version 8 and earlier, you
shouldn't add a selectorless `@Directive()` decorator.
You can either add `@Directive()` with a selector or

> add an explicit constructor to affected subclasses.

Reviewer: ```suggestion
move the Angular-specific features to affected subclasses.
```

To support Angular version 8, you should use a `@Directive()`
decorator with metadata or move the Angular-specific features

> to subclasses.

Reviewer: This section is a bit redundant with the paragraph above. Remove?

## What about applications using non-migrated libraries?

In this case, `ngcc` manages running the application.

Note that the migration never adds a definition in the schematic,
it only adds the `@Directive()` selector.

> 

Reviewer: ngcc doesn't really run applications - it's a build step that updates libraries on the fly to be Ivy-compatible. I think all we need for this question is:

```suggestion
`ngcc` should transform any non-migrated libraries to generate the proper code.
```
